src/main/scala/com/microsoft/ml/spark/lightgbm/dataset/DatasetAggregator.scala

```diff
@@ -0,0 +1,488 @@
+              weightChunkedArrayOpt: Option[FloatChunkedArray],
+              initScoreChunkedArrayOpt: Option[DoubleChunkedArray],
+              featuresChunkedArray: DoubleChunkedArray,
+              groupColumnValues: ListBuffer[Row],
+              numCols: Int): Unit
+}
+
+class DenseAggregatedColumns(chunkSize: Int) extends BaseDenseAggregatedColumns {
+  def incrementCount(rowCount: Long,
+                     initScoreCount: Long): Unit = {
+    this.rowCount += rowCount
+    this.initScoreCount += initScoreCount
+  }
+
+  /** Adds the rows to the internal data structure.
+    * @param labelsChunkedArray The column of label values.
+    * @param weightChunkedArrayOpt The optional column of weights, if specified.
+    * @param initScoreChunkedArrayOpt The optional column of initial scores, if specified.
+    * @param featuresChunkedArray The features vector.
+    * @param groupColumnValues The column of group values, if in ranking scenario.
+    */
+  def addRows(labelsChunkedArray: FloatChunkedArray,
+              weightChunkedArrayOpt: Option[FloatChunkedArray],
+              initScoreChunkedArrayOpt: Option[DoubleChunkedArray],
+              featuresChunkedArray: DoubleChunkedArray,
+              groupColumnValues: ListBuffer[Row],
+              numCols: Int): Unit = {
+    initializeRows(weightChunkedArrayOpt, initScoreChunkedArrayOpt, numCols)
+    // Coalesce to main arrays passed to dataset create
+    labelsChunkedArray.coalesceTo(this.labelsArray)
+    weightChunkedArrayOpt.foreach(_.coalesceTo(this.weightArrayOpt.get))
+    initScoreChunkedArrayOpt.foreach(_.coalesceTo(this.initScoreArrayOpt.get))
+    featuresChunkedArray.coalesceTo(this.featuresArray)
+    for (index <- 0 until groupColumnValues.length) {
+      groupColumnValuesArray(index) = groupColumnValues(index)
+    }
+  }
+
+  private def initializeRows(weightChunkedArrayOpt: Option[FloatChunkedArray],
+                             initScoreChunkedArrayOpt: Option[DoubleChunkedArray],
+                             numCols: Int): Unit = {
+    this.numCols = numCols
+    labelsArray = new FloatSwigArray(this.rowCount)
+    weightChunkedArrayOpt.foreach(weightChunkedArray => {
+      weightArrayOpt = Some(new FloatSwigArray(this.rowCount))
+    })
+    initScoreChunkedArrayOpt.foreach(initScoreChunkedArray => {
+      initScoreArrayOpt = Some(new DoubleSwigArray(this.initScoreCount))
+    })
+    featuresArray = new DoubleSwigArray(numCols * this.rowCount)
+    groupColumnValuesArray = new Array[Row](this.rowCount.toInt)
+  }
+}
+
+/** Defines class for aggregating rows to a single structure before creating the native LightGBMDataset.
+  * @param chunkSize The chunk size for the chunked arrays.
+  */
+class DenseSyncAggregatedColumns(chunkSize: Int) extends BaseDenseAggregatedColumns {
+  /**
+    * Variables for current thread to use in order to update common arrays in parallel
+    */
+  @volatile var threadRowStartIndex = 0L
```

Comment:
This is a great perspective and comment, thank you.  I agree I've made more things public than I should have.  I think this is a much bigger change outside the scope of this PR and I can make this change in a follow-up PR.  The main reason why this hasn't been much of a problem is that users tend to use just the spark estimator (LightGBMClassifier/LightGBMRegressor/LightGBMRanker) in scala, and in pyspark none of these methods are even available to users since they are not wrapped in python.

Reply:
I made many of the new classes private in the "old-new" PR https://github.com/Azure/mmlspark/pull/1066 since this newer/now older PR will be abandoned.